hollow/src/main/java/com/netflix/hollow/api/consumer/incubating/HollowConsumerV2.java

```diff
@@ -0,0 +1,515 @@
+package com.netflix.hollow.api.consumer.incubating;
+
+import static java.lang.System.arraycopy;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptySet;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+import com.netflix.hollow.api.client.HollowAPIFactory;
+import com.netflix.hollow.api.consumer.HollowConsumer;
+import com.netflix.hollow.api.consumer.fs.HollowFilesystemBlobRetriever;
+import com.netflix.hollow.api.custom.HollowAPI;
+import com.netflix.hollow.api.metrics.HollowConsumerMetrics;
+import com.netflix.hollow.api.metrics.HollowMetricsCollector;
+import com.netflix.hollow.core.read.filter.HollowFilterConfig;
+import com.netflix.hollow.core.util.DefaultHashCodeFinder;
+import com.netflix.hollow.core.util.HollowObjectHashCodeFinder;
+import java.io.File;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
+public class HollowConsumerV2<A extends HollowAPI> extends HollowConsumer {
+    public static Builder<HollowAPI> newBuilder() {
+        return new Builder<>();
+    }
+
+    private final HollowClientUpdaterV2<A> updaterV2;
+
+    protected HollowConsumerV2(Builder<A> builder) {
+        super(
+                builder.announcementWatcher.get(),
+                builder.updater,
+                builder.refreshExecutor.get(),
+                builder.metrics,
+                builder.refreshLock
+        );
+        this.updaterV2 = builder.updater;
+    }
+
+    public CompletableFuture<HollowConsumerV2<A>> initialLoad() {
+        return getInitialLoad().thenApply(l -> this);
+    }
+
+    @Override
+    public A getAPI() {
+        return updaterV2.getAPI();
+    }
+
+    /**
+     * Equivalent to calling {@link #getAPI()} and casting to the specified API.
+     *
+     * @param apiClass the class of the API
+     * @param <T> the type of the API
+     * @return the API which wraps the underlying dataset
+     *
+     * @deprecated use type parameter on consumer, e.g. {@code HollowConsumerV2<T>} then call {@link #getAPI()}
+     */
+    @Override
+    @Deprecated
+    public <T extends HollowAPI> T getAPI(Class<T> apiClass) {
+        return apiClass.cast(updater.getAPI());
+    }
+
+    public static class Builder<A extends HollowAPI> {
```

Comment:
Unlike `HollowConsumer`, the existing `HollowConsumer.Builder` is already parameterized. This would be a breaking change if applied to `HollowConsumer.Builder`